src/services/services.ts

```diff
@@ -899,6 +899,13 @@ namespace ts {
         sourceFile.scriptSnapshot = scriptSnapshot;
     }
 
+    export function overrideCreateupdateLanguageServiceSourceFile(

```

Comment:
We have two options:

1. in `LanguageServiceHost`:
```ts
export interface LanguageServiceHost {
    ...
    createSourceFile?(fileName: string, scriptSnapshot: IScriptSnapshot, scriptTarget: ScriptTarget, version: string, setNodeParents: boolean, scriptKind?: ScriptKind): SourceFile;
    updateSourceFile?(sourceFile: SourceFile, scriptSnapshot: IScriptSnapshot, version: string, textChangeRange: TextChangeRange, aggressiveChecks?: boolean): SourceFile;
    ....
}
```

Then in the language service, in places where we call `ts.createLanguageServiceSourceFile` or `ts.updateLanguageServiceSourceFile`, we can just check the host if it supports these APIs.

2. users can wrap the `DocumentRegistry` for the language service, and do the needed work in `acquireDocument` and `updateDocument`.